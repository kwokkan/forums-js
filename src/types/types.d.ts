
interface IForumSessionUser {
    id: number;
    name: string;
    joinedDate: number;
}

interface INextSession extends any {
    forumsUser?: IForumSessionUser;
}

interface INextProvider extends any {
    id: string;
    name: string;
}

declare module "next-auth" {
    interface INextAccount extends any {
        provider: string;
        id: string;
    }

    export interface IProviderOptions {
        providers: any[];
        callbacks?: IProviderOptionsCallback;
        events?: INextEventsOptions;
        pages?: INextPageOptions;
        debug?: boolean;
    }

    interface IProviderOptionsCallback {
        signIn?: (user: INextUser, account: INextAccount, profile: INextProfile) => Promise<bool>;
        redirect?: (url: string, baseUrl: string) => Promise<string>;
        session?: (session: INextSession, user: INextUser, sessionToken: string) => Promise<INextSession>;
        jwt?: (token: INextToken, user: INextUser, account: INextAccount, profile: INextProfile, isNewUser: boolean) => Promise<INextToken>;
    }

    interface INextEventsOptions {
        signIn?: (message: any) => Promise;
        signOut?: (message: any) => Promise;
        createUser?: (message: any) => Promise;
        linkAccount?: (message: any) => Promise;
        session?: (message: any) => Promise;
        error?: (message: any) => Promise;
    }

    interface INextPageOptions {
        signIn?: string;
        signOut?: string;
        error?: string;
        verifyRequest?: string;
        newUser?: string;
    }

    interface INextProfile extends any {
    }


    interface INextToken extends any {
        forumsUser?: IForumSessionUser;
    }

    interface INextUser extends any {
        name: string;
        forumsUser?: IForumSessionUser;
    }

    function handler(req: IncomingMessage, res: ServerResponse, options: IProviderOptions): void;

    export = handler;
}

declare module "next-auth/client" {
    export const Provider: any;

    export function getSession(context: { req: IncomingMessage }): Promise<INextSession>;

    export function providers<T>(context: GetServerSidePropsContext<T>): Promise<INextProvider[]>;

    export function signIn(provider?: string, args?: { callbackUrl?: string }): void;

    export function signOut(): void;

    export function useSession(): [INextSession | null, boolean];
}

declare module "next-auth/providers" {
    interface IProviderConfig {
        clientId: string,
        clientSecret?: string,
        scope?: string[] | string
    }

    export = {
        GitHub: (config: IProviderConfig) => any
    };
}
