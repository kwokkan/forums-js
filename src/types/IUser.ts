export interface IUser {
    id: number;
    name: string;
    joinedDate: Date;
}
