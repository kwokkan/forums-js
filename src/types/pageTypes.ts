import { GetServerSideProps } from "next";
import { ParsedUrlQuery } from "querystring";

export type GetTypedServerSideProps<TParams extends ParsedUrlQuery, TProps = { [key: string]: any }> = GetServerSideProps<TProps, TParams>;
