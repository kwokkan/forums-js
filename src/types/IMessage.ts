import { IUser } from "./IUser";

export interface IMessage {
    id: number;
    created: Date;
    content: string;
    user: IUser;
}
