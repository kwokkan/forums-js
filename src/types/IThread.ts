import { IMessage } from "./IMessage";

export interface IThread {
    id: number;
    created: number;
    name: string;
    messages: IMessage[];
}
