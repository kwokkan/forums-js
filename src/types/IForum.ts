import { IThread } from "./IThread";

export interface IForum {
    id: number;
    created: Date;
    name: string;
    description?: string;
    threads?: IThread[];
}
