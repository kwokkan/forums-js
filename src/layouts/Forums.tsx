import Link from "next/link";
import React from "react";
import { Content, Panel, PanelGroup } from "rsuite";
import { IForum } from "../types/IForum";

export interface IProps {
    forums: IForum[];
}

export function Forums(props: IProps) {
    return (
        <PanelGroup>
            {props.forums.map(x =>
                <Panel key={x.id} header={<h3><Link href={`/forums/${x.id}`}><a>{x.name}</a></Link></h3>} bordered>
                    {x.description &&
                        <Content>
                            <p>
                                {x.description}
                            </p>
                        </Content>
                    }
                </Panel>
            )}
        </PanelGroup>
    );
}
