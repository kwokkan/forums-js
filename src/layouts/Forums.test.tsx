import React from "react";
import { create } from "react-test-renderer";
import { IForum } from "../types/IForum";
import { Forums } from "./Forums";

test("Renders with no forums", () => {
    const forums: IForum[] = [];

    const tree = create(
        <Forums forums={forums} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

test("Renders with multiple forums", () => {
    const created = new Date(2020, 1, 1).getTime();
    const forums: IForum[] = [
        {
            id: 1,
            created: created,
            name: "First forum"
        },
        {
            id: 2,
            created: created,
            name: "Second forum"
        }
    ];

    const tree = create(
        <Forums forums={forums} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

test("Renders with description", () => {
    const created = new Date(2020, 1, 1).getTime();
    const forums: IForum[] = [
        {
            id: 1,
            created: created,
            name: "First forum",
            description: "Forum description."
        }
    ];

    const tree = create(
        <Forums forums={forums} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
