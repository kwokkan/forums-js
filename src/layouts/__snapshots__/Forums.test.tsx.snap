// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with description 1`] = `
<div
  className="rs-panel-group"
>
  <div
    className="rs-panel rs-panel-default rs-panel-bordered"
  >
    <div
      className="rs-panel-heading"
      onClick={[Function]}
      role="rowheader"
      tabIndex={-1}
    >
      <h3
        className="rs-panel-title"
      >
        <a
          href="/forums/1"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          First forum
        </a>
      </h3>
    </div>
    <div
      className="rs-panel-body"
    >
      <div
        className="rs-content"
      >
        <p>
          Forum description.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders with multiple forums 1`] = `
<div
  className="rs-panel-group"
>
  <div
    className="rs-panel rs-panel-default rs-panel-bordered"
  >
    <div
      className="rs-panel-heading"
      onClick={[Function]}
      role="rowheader"
      tabIndex={-1}
    >
      <h3
        className="rs-panel-title"
      >
        <a
          href="/forums/1"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          First forum
        </a>
      </h3>
    </div>
    <div
      className="rs-panel-body"
    />
  </div>
  <div
    className="rs-panel rs-panel-default rs-panel-bordered"
  >
    <div
      className="rs-panel-heading"
      onClick={[Function]}
      role="rowheader"
      tabIndex={-1}
    >
      <h3
        className="rs-panel-title"
      >
        <a
          href="/forums/2"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Second forum
        </a>
      </h3>
    </div>
    <div
      className="rs-panel-body"
    />
  </div>
</div>
`;

exports[`Renders with no forums 1`] = `
<div
  className="rs-panel-group"
/>
`;
